File: android/src/main/kotlin/com/pubmatic/sdk/flutter_openwrap_sdk/interstitial/POBInterstitialEventHandlerClient.kt
package com.pubmatic.sdk.flutter_openwrap_sdk.interstitial

import com.pubmatic.sdk.common.POBError
import com.pubmatic.sdk.common.log.POBLog
import com.pubmatic.sdk.flutter_openwrap_sdk.POBFlutterConstants
import com.pubmatic.sdk.flutter_openwrap_sdk.POBUtils
import com.pubmatic.sdk.openwrap.core.POBBid
import com.pubmatic.sdk.openwrap.interstitial.POBInterstitialEvent
import com.pubmatic.sdk.openwrap.interstitial.POBInterstitialEventListener
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * The OpenWrap interstitial custom event abstract class. Your interstitial custom event handler must
 * implement this class to communicate with ad server SDK.
 */
abstract class POBFLTInterstitialEvent : POBInterstitialEvent() {
  /**
   * The method is called by the client to pass the callbacks from the flutter event handler
   * back to the OW-SDK.
   *
   * @param methodName Method name to be called
   * @param call Method call
   * @param result Result of the method call used to pass data back to flutter part
   */
  abstract fun methodCall(methodName: String, call: MethodCall, result: MethodChannel.Result)
}

/**
 * Implementation of [POBInterstitialEvent] which is used as the event handler for OW-SDK in case of
 * header bidding.
 *
 * @param adId AdId used for storing the Ad object in the map
 * @param channel Instance of [MethodChannel] used to call flutter methods and pass data
 */
class POBInterstitialEventHandlerClient(
  private val adId: Int,
  private val channel: MethodChannel
) :
  POBFLTInterstitialEvent() {

  private var eventListener: POBInterstitialEventListener? = null

  override fun requestAd(bid: POBBid?) {
    eventListener?.let {
      val openWrapTargeting = it.bidsProvider?.targetingInfo
      channel.invokeMethod("requestAd", POBUtils.getArgumentMap(adId, openWrapTargeting))
    }
  }

  override fun methodCall(methodName: String, call: MethodCall, result: MethodChannel.Result) {
    when (methodName) {
      "onAdOpened" -> {
        eventListener?.onAdOpened()
        result.success(null)
      }

      "onAdClosed" -> {
        eventListener?.onAdClosed()
        result.success(null)
      }

      "onAdServerWin" -> {
        POBLog.info(TAG, "Ad Server won.")
        eventListener?.onAdServerWin()
        result.success(null)
      }

      "onAdExpired" -> {
        eventListener?.onAdExpired()
        result.success(null)
      }

      "onAdClick" -> {
        eventListener?.onAdClick()
        result.success(null)
      }

      "onAdLeftApplication" -> {
        eventListener?.onAdLeftApplication()
        result.success(null)
      }

      "onAdImpression" -> {
        POBLog.info(TAG, "GAM interstitial recorded the impression")
        eventListener?.onAdImpression()
        result.success(null)
      }

      "onFailedToLoad" -> {
        eventListener?.onFailedToLoad(
          POBError(
            call.argument<Int>(POBFlutterConstants.KEY_ERROR_CODE)!!,
            call.argument<String>(POBFlutterConstants.KEY_ERROR_MESSAGE)!!
          )
        )
        result.success(null)
      }

      "onFailedToShow" -> {
        eventListener?.onFailedToShow(
          POBError(
            call.argument<Int>(POBFlutterConstants.KEY_ERROR_CODE)!!,
            call.argument<String>(POBFlutterConstants.KEY_ERROR_MESSAGE)!!
          )
        )
        result.success(null)
      }

      "onOpenWrapPartnerWin" -> {
        POBLog.info(TAG, "OpenWrap Partner win.")
        eventListener?.onOpenWrapPartnerWin(call.argument<String>("bidId"))
        result.success(null)
      }

      else -> result.notImplemented()
    }
  }

  override fun show() {
    channel.invokeMethod("show", POBUtils.getArgumentMap(adId))
  }

  override fun destroy() {
    eventListener = null
  }

  override fun setEventListener(eventListener: POBInterstitialEventListener) {
    this.eventListener = eventListener
  }

  companion object {
    private const val TAG = "POBInterstitialEventHandlerClient"
  }
}

File: ios/Classes/POBInterstitialEventHandlerClient.swift
import OpenWrapSDK
import Flutter

/// The interstitial custom event delegate. It is used to inform ad server events back to the OpenWrap SDK
protocol POBFLTInterstitialEvent: POBInterstitialEvent {
    /// Method to route all the OpenWrapSDK global API calls
    /// - Parameters:
    ///     - name: method to invoke
    ///     -  call: FlutterMethodCall for argument & other details
    ///     - result: result block to invoke
    func invokeMethod(methodName: String, call: FlutterMethodCall, result: FlutterResult)
}

/// Wrapper class around `POBInterstitialEvent` to transfer the API calls
///
/// - Parameters:
/// - methodChannel: `MethodChannel` for native-dart communication
/// - delegate: instance of POBInterstitialEventDelegate
/// - adId: Int to uniquely identify ad request
/// - bid: Instance of POBBid

class POBInterstitialEventHandlerClient: NSObject, POBFLTInterstitialEvent {
    var methodChannel: FlutterMethodChannel?
    weak var eventDelegate: POBInterstitialEventDelegate?
    var adId: Int

    init(methodChannel: FlutterMethodChannel?, adId: Int) {
        self.methodChannel = methodChannel
        self.adId = adId
        super.init()
    }

    func requestAd(with bid: POBBid?) {
        let bidsProvider = eventDelegate?.bidsProvider()
        let targeting = bidsProvider?.targettingInfo()
        methodChannel?.invokeMethod("requestAd", arguments: POBUtils.convertCustomTargetingToMap(targeting: targeting,
                                                                                                 adId: adId))

    }

    func setDelegate(_ delegate: POBInterstitialEventDelegate) {
        self.eventDelegate = delegate
    }

    func show(from controller: UIViewController) {
        methodChannel?.invokeMethod("show", arguments: ["adId": adId])
    }

    deinit {
        methodChannel?.invokeMethod("destroy", arguments: ["adId": adId])
        methodChannel = nil
        eventDelegate = nil
    }

    /// The method is called by the client to pass the callbacks from the flutter event handler
    /// back to the OW-SDK.
    ///
    /// - Parameters:
    ///     - methodName: Method name to be called
    ///     - call: Method call
    ///     - result: Result of the method call used to pass data back to flutter part
    func invokeMethod(methodName: String, call: FlutterMethodCall, result: FlutterResult) {
        switch methodName {
        case "onOpenWrapPartnerWin":
            if let args = call.arguments as? [String: Any], let bidId = args["bidId"] as? String {
                eventDelegate?.openWrapPartnerDidWin(forBid: bidId)
            }
            result(nil)
        case "onAdServerWin":
            eventDelegate?.adServerDidWin()
            result(nil)
        case "onAdOpened":
            eventDelegate?.didPresentAd()
            result(nil)
        case "onAdClosed":
            eventDelegate?.didDismissAd()
            result(nil)
        case "onAdExpired":
            eventDelegate?.adDidExpireAd()
            result(nil)
        case "onAdClick":
            eventDelegate?.didClickAd()
            result(nil)
        case "onAdLeftApplication":
            eventDelegate?.willLeaveApp()
            result(nil)
        case "onAdImpression":
            eventDelegate?.adServerAdDidRecordImpression()
            result(nil)
        case "onFailedToLoad":
            let values = call.arguments as? [String: Any]
            eventDelegate?.failedToLoadWithError(POBUtils.error(from: values))
            result(nil)
        case "onFailedToShow":
            let values = call.arguments as? [String: Any]
            eventDelegate?.failedToShowWithError(POBUtils.error(from: values))
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
